### PR TITLE
Recovery of missing hits for Strip Hit Efficiency

### PR DESCRIPTION
**Pull request by DenkMybu**
#### PR description: 

These changes implement the recovery of missing hits for the mkFit trajectories. Without this, the hit efficiency is overestimated in single-sided layers. 

Output changes : no changes in the output format
Other PRs : None 
Indico page for results : https://indico.cern.ch/event/1271487/  (Update on hit efficiency)

#### PR validation:

Version comparisons checked (see indico page above), tested with/without changes

### CODE REVIEW

**Comment by slava77** [comment]
The May 23 slides have 
1. Sub-detector only
2. Adding sub-detector transitions

this is the latter, with transitions, right?

**Comment by mmusich** [comment]
@cmsbuild, please test

File: CalibTracker/SiStripHitEfficiency/python/SiStripHitEff_cff.py [hunk 1]
```diff
@@ -36,7 +36,8 @@
                        useFirstMeas = cms.untracked.bool(False),
                        useLastMeas = cms.untracked.bool(False),
                        # use or not all hits when some missing hits in the trajectory (bias), default is false
-                       useAllHitsFromTracksWithMissingHits = cms.untracked.bool(False)
+                       useAllHitsFromTracksWithMissingHits = cms.untracked.bool(False),
+                       doMissingHitsRecovery = cms.untracked.bool(True)
```
[the remark this answers is not in the document]
Comment (jlagram) [reply CalibTracker/SiStripHitEfficiency/python/SiStripHitEff_cff.py:40]: Not sure if it is worth (I am not used to this). Currently the parameter must only be added in the configuration file of the calibTree production.

I just realise that we should set useAllHitsFromTracksWithMissingHits to True when using doMissingHitsRecovery and mkFit trajectories.

[the remark this answers is not in the document]
Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/python/SiStripHitEff_cff.py:40]: then it's better to use a run-3 era modifier that flips the flag for these parameters at the same time.

[the remark this answers is not in the document]
Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/python/SiStripHitEff_cff.py]: I have added the run-3 era modifier in the latest commit

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 2]
```diff
@@ -369,15 +379,57 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
 
       // Check whether the trajectory has some missing hits
       bool hasMissingHits = false;
+      unsigned int previous_layer = 999;
+      vector<unsigned int> missedLayers;
+
       for (itm = TMeas.begin(); itm != TMeas.end(); itm++) {
```
Comment (tvami) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:385]: this should be a range-based loop

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 3]
```diff
@@ -369,15 +379,57 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
         auto theHit = (*itm).recHit();
+        unsigned int iidd = theHit->geographicalId().rawId();
+        unsigned int layer = ::checkLayer(iidd, tTopo);
+        int MissedLayer = int(layer + 1);
+        int DiffPreviousLayer = int(layer - previous_layer);
```
Comment (tvami) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:390]: according to the CMS naming conventions, these should be lowercase

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 4]
```diff
@@ -872,6 +1054,93 @@ void HitEff::endJob() {
...
+    float totTID = 0.0;
+    float totTEC = 0.0;
+
+    float totTIBrepro = 0.0;
+    float totTOBrepro = 0.0;
+    float totTIDrepro = 0.0;
+    float totTECrepro = 0.0;
+
+    edm::LogInfo("SiStripHitEfficiency:HitEff") << "Within TIB :" << std::endl;
+    for (int i = 0; i < 5; i++) {
```
Comment (tvami) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:1070]: the 5 is a magic number here

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 5]
```diff
@@ -872,6 +1054,93 @@ void HitEff::endJob() {
...
+          << " % of missing hit" << std::endl;
+      totTIBrepro += (missHitPerLayer[i] - hitRecoveryCounters[i]);
+    }
+    edm::LogInfo("SiStripHitEfficiency:HitEff")
+        << "TOTAL % of missing hits within TIB :" << (totTIB * 1.0 / totalNbHits) * 100 << "%" << std::endl;
+    edm::LogInfo("SiStripHitEfficiency:HitEff")
+        << "AFTER repropagation :" << (totTIBrepro * 1.0 / totalNbHits) * 100 << "%" << std::endl;
+
+    edm::LogInfo("SiStripHitEfficiency:HitEff") << "Within TOB :" << std::endl;
+    for (int i = 5; i < 11; i++) {
```
Comment (tvami) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:1087]: here too

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 6]
```diff
@@ -872,6 +1054,93 @@ void HitEff::endJob() {
...
+          << " % of missing hit" << std::endl;
+      totTOBrepro += (missHitPerLayer[i] - hitRecoveryCounters[i]);
+    }
+    edm::LogInfo("SiStripHitEfficiency:HitEff")
+        << "TOTAL % of missing hits within TOB :" << (totTOB * 1.0 / totalNbHits) * 100 << "%" << std::endl;
+    edm::LogInfo("SiStripHitEfficiency:HitEff")
+        << "AFTER repropagation :" << (totTOBrepro * 1.0 / totalNbHits) * 100 << "%" << std::endl;
+
+    edm::LogInfo("SiStripHitEfficiency:HitEff") << "Within TID :" << std::endl;
+    for (int i = 11; i < 14; i++) {
```
Comment (tvami) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:1104]: here too

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 7]
```diff
@@ -872,6 +1054,93 @@ void HitEff::endJob() {
...
+          << " % of missing hit" << std::endl;
+      totTIDrepro += (missHitPerLayer[i] - hitRecoveryCounters[i]);
+    }
+    edm::LogInfo("SiStripHitEfficiency:HitEff")
+        << "TOTAL % of missing hits within TID :" << (totTID * 1.0 / totalNbHits) * 100 << "%" << std::endl;
+    edm::LogInfo("SiStripHitEfficiency:HitEff")
+        << "AFTER repropagation :" << (totTIDrepro * 1.0 / totalNbHits) * 100 << "%" << std::endl;
+
+    edm::LogInfo("SiStripHitEfficiency:HitEff") << "Within TEC :" << std::endl;
+    for (int i = 14; i < 23; i++) {
```
Comment (tvami) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:1121]: here too

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 8]
```diff
@@ -435,7 +487,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+            } else if (tTopo->tecSide(iidd) == 2) {
+              tmpTmeas = layerMeasurements.measurements(*tecLayerpos, tsos, *thePropagator, *estimator);
+            }
+          }
+
+          if (!tmpTmeas.empty()) {
+            TrajectoryMeasurement TM_tmp(tmpTmeas.back());
+            unsigned int iidd_tmp = TM_tmp.recHit()->geographicalId().rawId();
+            if (iidd_tmp != 0) {
+              LogDebug("SiStripHitEfficiency:HitEff") << " hit actually being added to TM vector" << endl;
```
Comment (tvami) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:592]: in general the LogInfo and LogDebug dont need the `<< endl` for the ending

**Review by tvami — COMMENTED** [review]
Comments on magic numbers and for loops + other cosmetic comments

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 9]
```diff
@@ -104,6 +104,12 @@ HitEff::HitEff(const edm::ParameterSet& conf)
   useLastMeas_ = conf_.getUntrackedParameter<bool>("useLastMeas", false);
   useAllHitsFromTracksWithMissingHits_ =
       conf_.getUntrackedParameter<bool>("useAllHitsFromTracksWithMissingHits", false);
+  doMissingHitsRecovery_ = conf_.getUntrackedParameter<bool>("doMissingHitsRecovery", false);
+
+  for (int ilayer = 0; ilayer < 23; ilayer++) {
```
[the remark this answers is not in the document]
Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:109]: using the topology to get these loop boundaries will obfuscate the code as mush as using magic numbers.
I would propose, instead, to use these constants instead:

https://github.com/cms-sw/cmssw/blob/e0f88d0cd93d094405852b122ff61a2cbc9597a0/CalibTracker/SiStripHitEfficiency/interface/SiStripHitEfficiencyHelpers.h#L25-L33

[the remark this answers is not in the document]
Comment (tvami) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:109]: yes, that sounds great!

**Comment by mmusich** [comment]
please test

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 2 above)
Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:383]: for the record, the suggestion is

```diff
- for (itm = TMeas.begin(); itm != TMeas.end(); itm++) {
+ for(const auto& itm : TMeas){
```

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 10]
```diff
@@ -369,16 +377,56 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+          missedLayers.push_back(layer);
+        previous_layer = layer;
       }
 
       // Loop on each measurement and take it into consideration
       //--------------------------------------------------------
-
-      for (itm = TMeas.begin(); itm != TMeas.end(); itm++) {
+      unsigned int prev_TKlayers = 0;
+      for(const auto& itm : TMeas) {
```
Comment (mmusich) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:429]: I don't think using the range-based loop here is appropriate.

Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:429]: Sorry for the mistake I went too fast ! Not appropriate here indeed

Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc]: Do you know why changing it back causes an additional error ?

Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc]: there error in not here, it's in the above loop.

Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc]: see https://github.com/cms-sw/cmssw/pull/41970#discussion_r1235265411

Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc]: I can't compile anymore because my version is deprecated, is changing

"auto theHit = (*itm).recHit():"
to
"auto theHit = (itm).recHit()" 

solving this issue in your opinion ?

Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc]: yes it is.

**Comment by mmusich** [comment]
by the way, this PR now has too many commits, please squash them together after rebasing

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 8 above)
Comment (mmusich) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:588]: finally, please address also this comment @DenkMybu 
```suggestion
              LogDebug("SiStripHitEfficiency:HitEff") << " hit actually being added to TM vector";
```

**Comment by mmusich** [comment]
@DenkMybu if you are wondering whats the correct course of action, here's a quick recipe:
```
cmsrel CMSSW_13_2_X_2023-06-20-1100
cd CMSSW_13_2_X_2023-06-20-1100/src/
cmsenv
git cms-merge-topic 41970
scram b code-format
git add CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc
git commit -m "code-format"
git rebase -i CMSSW_13_2_X_2023-06-20-1100
# in interactive squash commits such that it looks like this [1]
git push my-cmssw +HEAD:mkfit_missinghits
```


[1]
Screenshot from 2023-06-20 16-04-01

**Comment by DenkMybu** [comment]
Everything works fine except for the last git push, after squashing, error states 'my-cmssw- does not appear to be a git repository

**Comment by francescobrivio** [comment]
+1

 - tests are clean
 - only one minor comment (on `LogDebug` syntax) from the review is missed (see [here](https://github.com/cms-sw/cmssw/pull/41970#discussion_r1235288790)), but since I understood a followup PR to integrate this changes in PCL is needed, that can be addressed later

**Comment by mmusich** [comment]
@cms-sw/orp-l2 is there something missing here?

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 11]
```diff
@@ -369,15 +377,55 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
-        auto theHit = (*itm).recHit();
+      unsigned int previous_layer = 999;
+      vector<unsigned int> missedLayers;
+
+      for (const auto& itm : TMeas) {
+        auto theHit = itm.recHit();
+        unsigned int iidd = theHit->geographicalId().rawId();
+        unsigned int layer = ::checkLayer(iidd, tTopo);
+        int missedLayer = int(layer + 1);
+        int diffPreviousLayer = int(layer - previous_layer);
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:388]: As such this cannot work: making a difference of two `unsigned int` can never return a negative number, see L391 below.
At the very minimum this could be
```suggestion
        int diffPreviousLayer = int(layer) - int(previous_layer);
```
but probably redefining `layer` and `previous_layer` as `int` since the beginning could avoid some gymnastic

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 12]
```diff
@@ -369,15 +377,55 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+        if (doMissingHitsRecovery_) {
+          //Layers from TIB + TOB
+          if (diffPreviousLayer == -2 && missedLayer > k_LayersStart && missedLayer < k_LayersAtTOBEnd) {
+            missHitPerLayer[missedLayer] += 1;
+            hasMissingHits = true;
+          }
+          //Layers from TID
+          if (diffPreviousLayer == -2 && (missedLayer == 12 || missedLayer == 13)) {
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:396]: Not obvious whether `missedLayer == 12` or `missedLayer == 13` can enter in the previous condition.
Perhaps an `else if` (here and at [L401](https://github.com/cms-sw/cmssw/pull/41970/files#diff-2605c956edf7b5d5b9ae0863fc9f246940b9ba7274eb40a0747fcd088fe7a3c2R401)) can help clarifying

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 13]
```diff
@@ -369,15 +377,55 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+            hasMissingHits = true;
+          }
+          //Layers from TEC
+          if (diffPreviousLayer == -2 && missedLayer > k_LayersAtTIDEnd && missedLayer < k_LayersAtTECEnd) {
+            missHitPerLayer[missedLayer] += 1;
+            hasMissingHits = true;
+          }
+
+          //##### TID Layer 11 (transition TID -> TIB)
+          if ((int(layer) > k_LayersStart && int(layer) <= k_LayersAtTIBEnd) && (previous_layer == 12)) {
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:407]: Previous conditions were on `missedLayer,` these ones one `layer`: as a reviewer it is hard to follow, I imagine the same could be for a future developer who did not participate to this update to the code...

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 14]
```diff
@@ -435,7 +483,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
           //only add one TM for the single surface and the other will be added in the next iteration
           TMs.push_back(TrajectoryAtInvalidHit(*itm, tTopo, tkgeom, propagator));
         }
+        bool missingHitAdded = false;
+
+        vector<TrajectoryMeasurement> tmpTmeas;
+        unsigned int misLayer = TKlayers + 1;
+        //Use bool doMissingHitsRecovery to add possible missing hits based on actual/previous hit
+        if (doMissingHitsRecovery_) {
+          if (int(TKlayers - prev_TKlayers) == -2) {
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:492]: Again, the difference between two `unsigned int` can never become "-2"

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 15]
```diff
@@ -435,7 +483,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+            const TrajectoryStateOnSurface tsos = itm->updatedState();
+            std::vector<DetLayer::DetWithState> compatDets = detlayer->compatibleDets(tsos, *thePropagator, *estimator);
+
+            if (misLayer > k_LayersAtTIDEnd && misLayer < k_LayersAtTECEnd) {  //TEC
+              std::vector<ForwardDetLayer const*> negTECLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->negTecLayers();
+              std::vector<ForwardDetLayer const*> posTECLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->posTecLayers();
+              const DetLayer* tecLayerneg = negTECLayers[misLayer - 14];
+              const DetLayer* tecLayerpos = posTECLayers[misLayer - 14];
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:504]:
```suggestion
              const DetLayer* tecLayerneg = negTECLayers[misLayer - k_LayersAtTIDEnd -1];
              const DetLayer* tecLayerpos = posTECLayers[misLayer - k_LayersAtTIDEnd -1];
```

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 16]
```diff
@@ -435,7 +483,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+
+            if (misLayer > k_LayersStart && misLayer < k_LayersAtTOBEnd) {  // Barrel
+
+              std::vector<BarrelDetLayer const*> barrelTIBLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->tibLayers();
+              std::vector<BarrelDetLayer const*> barrelTOBLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->tobLayers();
+
+              if (misLayer > k_LayersStart && misLayer <= k_LayersAtTIBEnd) {
+                const DetLayer* tibLayer = barrelTIBLayers[misLayer - 1];
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:536]:
```suggestion
                const DetLayer* tibLayer = barrelTIBLayers[misLayer - k_LayersStart - 1];
```

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 17]
```diff
@@ -435,7 +483,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+                tmpTmeas = layerMeasurements.measurements(*tibLayer, tsos, *thePropagator, *estimator);
+              }
+              if (misLayer > k_LayersAtTIBEnd && misLayer < k_LayersAtTOBEnd) {
+                const DetLayer* tobLayer = barrelTOBLayers[misLayer - 5];
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:540]:
```suggestion
                const DetLayer* tobLayer = barrelTOBLayers[misLayer - k_LayersAtTIBEnd -1];
```

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 18]
```diff
@@ -435,7 +483,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+              std::vector<BarrelDetLayer const*> barrelTIBLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->tibLayers();
+              std::vector<BarrelDetLayer const*> barrelTOBLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->tobLayers();
+
+              if (misLayer > k_LayersStart && misLayer <= k_LayersAtTIBEnd) {
+                const DetLayer* tibLayer = barrelTIBLayers[misLayer - 1];
+                tmpTmeas = layerMeasurements.measurements(*tibLayer, tsos, *thePropagator, *estimator);
+              }
+              if (misLayer > k_LayersAtTIBEnd && misLayer < k_LayersAtTOBEnd) {
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:539]:
```suggestion
              else if (misLayer > k_LayersAtTIBEnd && misLayer < k_LayersAtTOBEnd) {
```

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 8 above)
Comment (perrotta) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:588]: This comment still needs to be addressed

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 12 above)
Comment (perrotta) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:396]:
```suggestion
          else if (diffPreviousLayer == -2 && missedLayer > k_LayersAtTOBEnd +1 && missedLayer <= k_LayersAtTIDEnd) {
```
(by the way, layer = 11 is not in TID?)

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 19]
```diff
@@ -369,15 +377,55 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+            missHitPerLayer[missedLayer] += 1;
+            hasMissingHits = true;
+          }
+          //Layers from TID
+          if (diffPreviousLayer == -2 && (missedLayer == 12 || missedLayer == 13)) {
+            missHitPerLayer[missedLayer] += 1;
+            hasMissingHits = true;
+          }
+          //Layers from TEC
+          if (diffPreviousLayer == -2 && missedLayer > k_LayersAtTIDEnd && missedLayer < k_LayersAtTECEnd) {
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:401]:
```suggestion
          else if (diffPreviousLayer == -2 && missedLayer > k_LayersAtTIDEnd && missedLayer < k_LayersAtTECEnd) {
```
(again: shouldn't  k_LayersAtTECEnd be also considered, i.e. "<" -> "<=")?

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 20]
```diff
@@ -435,7 +483,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+              const DetLayer* tecLayerneg = negTECLayers[misLayer - 14];
+              const DetLayer* tecLayerpos = posTECLayers[misLayer - 14];
+              if (tTopo->tecSide(iidd) == 1) {
+                tmpTmeas = layerMeasurements.measurements(*tecLayerneg, tsos, *thePropagator, *estimator);
+              } else if (tTopo->tecSide(iidd) == 2) {
+                tmpTmeas = layerMeasurements.measurements(*tecLayerpos, tsos, *thePropagator, *estimator);
+              }
+            }
+
+            else if (misLayer >= 12 && misLayer <= 13) {  // This is for outer TID layer (13) and layer 12
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:512]:
```suggestion
            else if (misLayer == 12 || misLayer == 13) {  // This is for outer TID layer (13) and layer 12
```
This is more intuitive. Even better if written in term of the parameters `k_LayersAtTOBEnd` and `k_LayersAtTIDEnd`.

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc [hunk 21]
```diff
@@ -435,7 +483,136 @@ void HitEff::analyze(const edm::Event& e, const edm::EventSetup& es) {
...
+
+              std::vector<ForwardDetLayer const*> negTIDLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->negTidLayers();
+              std::vector<ForwardDetLayer const*> posTIDLayers =
+                  measurementTrackerHandle->geometricSearchTracker()->posTidLayers();
+              const DetLayer* tidLayerneg = negTIDLayers[misLayer - 11];
+              const DetLayer* tidLayerpos = posTIDLayers[misLayer - 11];
```
Comment (perrotta) [inline CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:519]:
```suggestion
              const DetLayer* tidLayerneg = negTIDLayers[misLayer - k_LayersAtTOBEnd -1];
              const DetLayer* tidLayerpos = posTIDLayers[misLayer - k_LayersAtTOBEnd -1];
```

**Review by perrotta — COMMENTED** [review]
I see quite several issues with the differences among unsigned integers: was this code ever tested for all cases?

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 11 above)
Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:388]: Ok, I redefined those as ints so the rest will be changed aswell. Thanks

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 21 above)
Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:519]: Should these not be +1 instead of -1 ?

const DetLayer* tidLayerneg = negTIDLayers[misLayer - k_LayersAtTOBEnd + 1];

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 13 above)
Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:407]: I thought it would be better since the goal is to know if given layer is in TIB, while previous hit was before-last TID layer. changing it to "missedLayer" would add confusion would it not ?

File: CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc (hunk 12 above)
Comment (DenkMybu) [reply CalibTracker/SiStripHitEfficiency/plugins/HitEff.cc:396]: Yes, but with the condition "diffPreviousLayer == -2", the layer = 11 will never enter in this case

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
I tried to compare the hit efficiency:
  * using vanilla code
  * using the modifications in 8ddc8a319256bc77cb902da46ebbf6639ed1648b
  * using the modifications in 8ddc8a319256bc77cb902da46ebbf6639ed1648b + c8dcc34775ab9cfe5a3bdf87c952535ebb28c7d6 
  
  and it looks like the latter two deliver the exact same effciency:
  
Screenshot from 2023-06-26 13-38-03

is it expected?

**Comment by perrotta** [comment]
> is it expected?

I think so. Even if it is a bit counterintuitive to me, and certainly error prone, "-2" as uint is 4294967294 , but I've checked now that `int(4294967294) = -2` (as it takes also into account the sign bit).
Therefore, my worries about differences of unsigned integers were probably not correct, at least for the code in this PR. Still, I'd fill a bit unconfortable in trusting in differences between unsigned integers, and the updates applied after the last commit clarify thinngs a bit, I guess.

**Comment by mmusich** [comment]
@perrotta 

> I think so. Even if it is a bit counterintuitive to me, and certainly error prone, "-2" as uint is 4294967294 , but I've checked now that int(4294967294) = -2 (as it takes also into account the sign bit).

thank you for the check.

@cms-sw/alca-l2 please sign at earliest convenience, as already reported, this needs to be backported (and integrated in the PCL as well).

**Comment by francescobrivio** [comment]
+1

**Comment by perrotta** [comment]
+1
